quixstreams/rowconsumer/buffering.py

```diff
@@ -0,0 +1,403 @@
+        """
+        A group of individual `PartitionBuffer`s by partition.
+
+        :param partition: partition number.
+        :param max_size: the maximum size of the underlying `PartitionBuffer`s.
+        """
+        self._partition = partition
+        self._max_size = max_size
+        self._partition_buffers: dict[str, PartitionBuffer] = {}
+        self._partition_buffers_values = self._partition_buffers.values()
+
+    @property
+    def partition(self) -> int:
+        return self._partition
+
+    def assign_partition(self, topic: str):
+        """
+        Add a new partition to the buffer group.
+
+        :param topic: topic name.
+        """
+        if topic in self._partition_buffers:
+            return
+        # New partition, need to create a new buffer and add it to the heap
+        partition_buffer = PartitionBuffer(
+            partition=self._partition, topic=topic, max_size=self._max_size
+        )
+        self._partition_buffers[topic] = partition_buffer
+
+    def revoke_partition(self, topic: str):
+        """
+        Remove partition from the buffer group.
+
+        :param topic: topic name.
+        """
+        self._partition_buffers.pop(topic)
+
+    def set_high_watermarks(self, offsets: dict[str, int]):
+        """
+        Set high watermarks for assigned partitions.
+
+        :param offsets: a mapping of {<topic>: <offset>} with the high watermarks
+            for this group.
+        """
+        for topic, watermark in offsets.items():
+            buffer = self._partition_buffers[topic]
+            buffer.set_high_watermark(watermark)
+
+    def append(self, message: SuccessfulConfluentKafkaMessageProto):
+        """
+        Add a new message to the buffer group.
+
+        :param message: a successful Kafka message.
+        """
+
+        partition_buffer = self._partition_buffers[message.topic()]
+        partition_buffer.append(message=message)
+
+    def pop(self) -> Optional[SuccessfulConfluentKafkaMessageProto]:
+        """
+        Pop a message from the partition buffer with the smallest next_timestamp
+        and return it.
+
+        How it works:
+
+        - When the group has multiple partitions assigned, it will pop the message
+            from the partition with the smallest next timestamp.
+        - When there's only one partition in the group, it will pop a message
+            from this partition buffer.
+        """
+
+        buffers = self._partition_buffers_values
+        if len(buffers) == 1:
+            # Use iter() here to avoid creating a new list each time
+            buffer = next(iter(buffers))
+            return buffer.popleft()
+        elif len(buffers) > 1:
+            # There's more than one partition in the group and one of them is empty
+            # but not idle.
+            # Wait until new messages are fetched or the highwater is set.
+            for buffer in buffers:
+                if buffer.empty() and buffer.idle() != True:
```

Comment:
slightly more idiomatic
```suggestion
                if buffer.empty() and buffer.idle() is not True:
```